### PR TITLE
A very preliminary draft

### PR DESCRIPTION
**Pull request by wangkuiyi**

### CODE REVIEW

File: paddle/framework/framework.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+x.estimated = false                 # prevents from update.
+```
+
+Note that the variable has the following methods:
+
+```cpp
+class Variable {
+ public:
+  bool Estimated() const;
+  bool SetEstimated(bool);
```
Comment (jacquesqiao) [inline paddle/framework/framework.md]: what is the meaning of this return bool?

File: paddle/framework/framework.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+### Scope
+
+In programming languages, variables belong to scopes.  Scopes enable the release of local variables when a stack frame pops.
+
+A neural network is arguably equivalent to a program.  The *recurrent operator*, `paddle::operator::Recurrent`, is like `for` loop, and the conditional operator, `paddle::operator::Conditional`, is like `if/switch`.  They can have sub-networks as attributes, and `paddle::operator::Recurrent/Conditional::Run` creates a local scope before executing the sub-network.  At inference time, once `paddle::operator::Recurrent/Conditional::Run` frees the scope before completes.  At training time, it is the corresponding gradient operations' `paddle::operator::RecurrentGrad/ConditionalGrad::Run` who free the local scope.
+
+The global and nested local scopes form a hierarchy.  The following Python functions make it convenient to program scopes:
+
+1. `paddle.scope.current()` returns the current scope, which defaults to
```
Comment (jacquesqiao) [inline paddle/framework/framework.md]: is this line `which defaults to` not complete？

File: paddle/framework/framework.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  void Run(const Context& ctx) {
+    auto subscope = ctx.scope_.CreateScope("");
+    step_net_.Run(Context{subscope, ctx.places_, ctx.training_});
+    if (!ctx.training_) {
+      ctx.scope_.DeleteScope(subscope);
+    }
+};
+```
+
+Another example is that the Gemm operator need to create a tensor on `Context::places_[0]` and assign the tensor to its output variable:
```
Comment (QiJune) [inline paddle/framework/framework.md]: What about data parallelism on multi-GPUs? I think that the places_[0] should not be used inside a operator's run method.

File: paddle/framework/framework.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+```cpp
+class Gemm {
+ public:
+  void Run(const Context& ctx) {
+    if (paddle::platform::IsGPUPlace(ctx.places_[0])) {
+      cuDNNGemm(
+         Output(0).mutable_data<float>(ctx.places_[0], DerivedSizeFromInputs()),
+         ...);
```
Comment (QiJune) [inline paddle/framework/framework.md]: cudnnGemm needs to acquire a cudnnHandle to finish computation.
And nearly all the computation in Run method need to acquire a eigen device.
Please refer to #2648

File: paddle/framework/framework.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+and can be defined as
+
+```cpp
+struct Context {
+  Scope* scope_;
+  std::vector<Place> places_; // a network might run on multiple devices.
+  bool training_;
+};
+```
```
Comment (QiJune) [inline paddle/framework/framework.md]: I think that the context of Operator's Run method  is different from Net's Run method.
The Net can run on multi-devices, but the operator can only run on a specific device.
So, the operator maybe need a ```OpContext``` to Run.

File: paddle/framework/framework.md (hunk 3 above)
Comment (typhoonzero) [inline paddle/framework/framework.md]: Anyway, we must specify one place to aggregate data from multiple GPUs.

File: paddle/framework/framework.md (hunk 1 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: I think we can make it just void.

File: paddle/framework/framework.md (hunk 2 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: I meant that it defaults to the value returned by the next (second) line. I will rewrite it to make it clear.

File: paddle/framework/framework.md (hunk 5 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: To be frank I am not even sure that a net should be able to run on multiple devices. It seems that we can have multiple copies of a net, each running on a specific device, and we can do gradient aggregation using NCCL.  In this design, it seems that both net and operators need just one device.

File: paddle/framework/framework.md (hunk 3 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: @reyoung suggested defining the aggregation as an operator. I am not sure. It seems much easier if a net runs on a single device as in https://github.com/PaddlePaddle/Paddle/pull/2696#discussion_r125566396.

File: paddle/framework/framework.md (hunk 4 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: Good point. I am reading and reviewing your design.

File: paddle/framework/framework.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  }
+  return unactivated, act(unactivated, cap)
+}
+```
+
+Note that operators might call other operators.  In above example, `gemm` calls `act`.
+
+### Gradient Operators
+
+Each operator has a corresponding gradient operator that defines the gradient computation.
```
Comment (reyoung) [inline paddle/framework/framework.md]: One forward op is not only corresponding to one gradient operator but gradient operators.

For example, `FcOp` has three inputs. Not all of them needs gradient in practice. It could implement each input corresponding to one gradient operator, so `FcOp` could have three gradient ops.

More generally, Caffe2 and MXNet return an array of ops as one op's gradient ops.

File: paddle/framework/framework.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+Some Python API proposals:
+
+```python
+x = paddle.variable.new("x") # create a variable of not yet known type
+x = paddle.variable.tensor("x") # create a tensor typed varible without value
+x = paddle.variable.int("x", 10) # create an unnamed int varaible and set to 10
+x = paddle.variable.operation("x", paddle.operator.gemm(...)) # an operation
+x = paddle.variable.tensor("x", 
+      numpy.random.randn(200, 100), # set the value of the tensor
+      estimated = true)             # will be updated by the backward algorithm.
```
Comment (reyoung) [inline paddle/framework/framework.md]: `estimated` is not very straight forward name. Maybe `need_backward` or `requires_grad` is better? Not all variable that need gradient will be updated, only parameters do.

File: paddle/framework/framework.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+  };
+  Type type_;
+};
+```
+
+We can add more Place implementations, like FPGAPlace and XeonPhiPlace, in the future.
+
+### Gradient Operators
+
+A gradient operator should be build and linked only if we are building a binary that supports training.  If we are building an "inference-only" binary, we shouldn't link gradient operators.
```
Comment (reyoung) [inline paddle/framework/framework.md]: Gradient Operator could also be a forward operator. 

For example, operator `mul`'s gradient operator is also `mul`.

File: paddle/framework/framework.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+- tensor
+- operator
+- scope
+
+Some Python API proposals:
+
+```python
+x = paddle.variable.new("x") # create a variable of not yet known type
+x = paddle.variable.tensor("x") # create a tensor typed varible without value
+x = paddle.variable.int("x", 10) # create an unnamed int varaible and set to 10
```
Comment (reyoung) [inline paddle/framework/framework.md]: Is that variable `unnamed` or typo? Maybe `x = paddle.variable.int(10)` ?

Comment (reyoung) [inline paddle/framework/framework.md]: I think maybe an independent Scope could represent all attribute variables?

Like [here](https://github.com/PaddlePaddle/Paddle/issues/2743#issuecomment-313033202) I proposed.

File: paddle/framework/framework.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+### `OperatorProto`
+
+We'd like to generate operator creators automatically from C++ code so to keep Python code always updated.  So we need to describe each C++ class in a protobuf message `OperatorProto`. We also need to fill in an `OperatorProto` message for each operator class and expose these messages to Python function `paddle.framework.create_operation_creators`.  We call this filling and exposing mechanism *operator registration*.
+
+
+## Operators, Layers, Variables, Scope
+
+### Operators as Functions
+
+An operator is intrinsically a function, which has inputs and outputs.  For example, the functional representation of the GEMM operator is
```
Comment (reyoung) [inline paddle/framework/framework.md]: Maybe `InferShape` is also need?

File: paddle/framework/framework.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+gemm(X, W, scale, act=LeRU) {
+  unactivated = scale * X * W
+  if !act {
+    return unactivated
+  }
+  return unactivated, act(unactivated, cap)
+}
+```
+
+Note that operators might call other operators.  In above example, `gemm` calls `act`.
```
Comment (reyoung) [inline paddle/framework/framework.md]: Maybe a better example for an Op call other Ops is `RnnOp`. :-) Because normal gemm doesn't contain activation.

File: paddle/framework/framework.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+- half
+- string
+- tensor
+- operator
+- scope
+
+Some Python API proposals:
+
+```python
+x = paddle.variable.new("x") # create a variable of not yet known type
```
Comment (reyoung) [inline paddle/framework/framework.md]: This APIs could be unified by one `paddle.variable.new`?

```python
x_var = paddle.variable.new("x", 10)

y = paddle.variable.Tensor()
y_var = paddle.variable.new("y", y)

z = paddle.variable.new("z", "A string")

no_name = paddle.variable.new(name=None, val=100)  # actually I think each variable should have a name. It might have some independent scope to hold them.
```

File: paddle/framework/framework.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+x = paddle.variable.tensor("x") # create a tensor typed varible without value
+x = paddle.variable.int("x", 10) # create an unnamed int varaible and set to 10
+x = paddle.variable.operation("x", paddle.operator.gemm(...)) # an operation
```
Comment (reyoung) [inline paddle/framework/framework.md]: Also, why a variable can be an operation? Is that need for RnnOp?

File: paddle/framework/framework.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+ public:
+  bool Estimated() const;
+  bool SetEstimated(bool);
+  template <typename T> const T& Get() const;
+  template <typename T> T* GetMutable();
+  template <typename T> bool IsType(T) const;
+};
+```
+
+`Get` and `GetMutable` implements *lazy memory allocation*, as described in the [Variable design doc](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.md).
```
Comment (reyoung) [inline paddle/framework/framework.md]: Maybe only `GetMutable` implements *lazy memory allocation*. `Get` should raise an error if that variable is not created?

File: paddle/framework/framework.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  
+  Scope* CreateScope(StringPiece name); // Finds or creates a sub-scope.
+  void DeleteScope(Scope*);             // Delete a sub-scope or raise exception
+  
+ private:
+  std::map<std::string /*name*/, std::unique_ptr<Variable> > vars_;
+  std::shared_ptr<Scope> parent_;
+  std::vector<Scope*> children_;
+  
+  Mutex mutex_;  // Make this class thread-safe.
```
Comment (reyoung) [inline paddle/framework/framework.md]: It seems reasonable, maybe I should update Scope design and add mutex today.

File: paddle/framework/framework.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+};
+```
+
+## Execution and Context
+
+A neural network is a program.  Training or inference is to execute it.  The runtime environment of execution is known as a *context*:
+
+1. a scope,
+1. device(s), or places,
```
Comment (reyoung) [inline paddle/framework/framework.md]: Maybe the only device for GPU is not enough. It could be a `DeviceContext` for each GPU which holds

* computation stream
* handles for cuDNN/cuBLAS etc.

File: paddle/framework/framework.md (hunk 5 above)
Comment (reyoung) [inline paddle/framework/framework.md]: > it seems that both net and operators need just one device.

Is there a higher level concept to run `NCCL` in C++ or just let Python run `NCCL`? If that concept in CPP, it seems that is also a Network.

```cpp
class MultiDeviceNetwork {
private:
  // holds networks on each device.
  vector<Network> networks_;
}
```

However, I suggest that we should only concern about the single device in basic `Network`. It is easy to change a single device Network into multiple device `Network` by using NCCL.

File: paddle/framework/framework.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+  Variable* FindVar(std::string name); // Find in the hierrachy or return null.
+  Variable* CreateVar(StringPiece name, Variable*); // Find or create.
+  
+  Scope* CreateScope(StringPiece name); // Finds or creates a sub-scope.
+  void DeleteScope(Scope*);             // Delete a sub-scope or raise exception
+  
+ private:
+  std::map<std::string /*name*/, std::unique_ptr<Variable> > vars_;
+  std::shared_ptr<Scope> parent_;
+  std::vector<Scope*> children_;
```
Comment (jacquesqiao) [inline paddle/framework/framework.md]: Why we need children_ of a scope?

Comment (wangkuiyi) [inline paddle/framework/framework.md]: Debug printing is the only usage in my mind.

File: paddle/framework/framework.md (hunk 5 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: Up to what I understand now, I'd avoid running a network on multiple devices, because the maintenance of multiple CUDA streams means a very complicated CUDAContext or OpKernelContext or something like that.

And it seems that we can make use of multiple devices using data parallelism, which requires only running a net on a single device.

File: paddle/framework/framework.md (hunk 3 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: I agree with @QiJune on https://github.com/PaddlePaddle/Paddle/pull/2696#discussion_r125546577 -- using data parallelism and only `places_[0]` would be enough.

File: paddle/framework/framework.md (hunk 9 above)
Comment (wangkuiyi) [inline paddle/framework/framework.md]: I don't see that attribute is so special and cannot be represented by a Variable.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献文档。由于文档已迁移至[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo，因此关闭您的PR，欢迎您向[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) Repo贡献文档。
Thanks for contributing to PaddlePaddle! Since documents have been moved to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo, we close this PR. Welcome to contribute to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo.
